### ISSUE TITLE
[Feature Request] Titles

### ISSUE THREAD

**Issue by BlueFalconHD** [junegunn/fzf#3022]
## Problem
While creating a simple password manager with [another CLI tool](https://github.com/charmbracelet/skate) and FZF, I have come across this issue:
<img width="629" alt="image" src="https://user-images.githubusercontent.com/72631767/197588567-ec532334-56eb-41a4-93d1-f23e9f792f62.png">
The problem is that I add a header with the purpose of identifying that specific password manager against others. While it is still visible, it clutters the search window and would look much better sitting above it.

## Solution 
A simple solution to this could be to add a new flag for the header or add a new option that adds text in the top or bottom border.
<img width="813" alt="image" src="https://user-images.githubusercontent.com/72631767/197589057-9fdff36c-f4b5-4624-a7bb-36b92e5c9430.png">

### A simple idea for how this could work
```bash
skate get "$(skate list -krb | fzf --height 100% \
    --margin 30% \
    --border sharp \
    --layout reverse-list \
    --title "Passwords")" | clipcopy
```
This script fetches the names of all my passwords, feeds it into FZF and displays it like so:
<img width="827" alt="image" src="https://user-images.githubusercontent.com/72631767/197592030-42fd6af9-4526-4c6b-8f42-7525ff3fbf23.png">

**Comment by junegunn** [comment]
You might find `--header-first` useful.

```sh
fzf --header-first --header $'foo bar\n───────\n' --layout=reverse --border
```

Having said that, I'm not against having the option. But there are a few design choices we have to make.
* How are we going to display it when we don't use `--border`, or something like `--border=vertical` is used?
* Position of the title on the border: on the left or at the center? Should it be customizable?
* Styling of the title (e.g. Do we put spaces around it or not? Should it be customizable?)
    * <img width="561" alt="image" src="https://user-images.githubusercontent.com/700826/198170470-6890d114-1ee7-4f9a-91b3-490ba5f335aa.png">
    * <img width="561" alt="image" src="https://user-images.githubusercontent.com/700826/198170503-1e4619c0-29bf-43a3-9956-4b1963aa3c77.png">
    * <img width="561" alt="image" src="https://user-images.githubusercontent.com/700826/198170525-bdd80e59-bbda-4806-9274-88b61ce118af.png">
    * <img width="561" alt="image" src="https://user-images.githubusercontent.com/700826/198170587-50b035f2-7e37-4437-9fc8-350b8d114fa1.png">
    * <img width="561" alt="image" src="https://user-images.githubusercontent.com/700826/198170798-4fdee1d7-4a99-499e-a91f-245f05c49f2f.png">
    * <img width="561" alt="image" src="https://user-images.githubusercontent.com/700826/198170955-11bb30b5-9a8d-4aeb-898d-120a40d63d2b.png">
* Do we also need another set of options for the title for the preview window?

**Comment by BlueFalconHD** [comment]
I think full customization would be great. Some thoughts I have:
- For spaces around, just add a note somewhere that they can add them to their title.
- When no border is used, the title would be in the place that the border would be if it was shown.
- Alignment options would be great.
- A separate set of options for the preview window might be something that could be done in the future if people use the title option (if it ends up getting added) but doesn't seem that high priority right now.

Obligatory mention @junegunn

**Comment by junegunn** [comment]
Thanks for the feedback. I've decided to add `--border-label` and `--border-label-pos` for the following reasons.

* If we show the "title" without `--border`, there will be no clear visual separation between the title and the list of items, especially when the title is left-aligned, so I don't think users will appreciate it much without `--border`.
    * ```
        title
        5
        4
        3
        2
      > 1
        100/100
      >
      ```
* If we're going to associate this new "title" option to `--border`, we should give it a more explicit name, i.e.`--border-label=LABEL`; it's a bit wordy, but it clearly describes what it actually is. It expresses our intention that the option should be used with `--border`.
* Label is only shown when `--border` is one of `rounded`, `sharp`,  `horizontal`, `top`, or `bottom`. It will not be shown on `vertical`, `left`, or `right`.
* Another option called `--border-label-pos=INTEGER` will be added. A positive integer will left-align the label, negative will right-align it. `0`, the default, will put the label at the center of the border line.
* The color and the style of the label are customized via `--color label:...`, but it will support custom ANSI colors sequences like `--prompt`.

## Demo

```sh
# ANSI color codes are supported
# (with https://github.com/busyloop/lolcat)
label=$(curl -s http://metaphorpsum.com/sentences/1 | lolcat -f)

# Border label at the center
fzf --height=10 --border-label="╢ $label ╟" --border --color=label:italic:black

# Left-aligned (positive integer)
fzf --height=10 --border-label="╢ $label ╟" --border=top --border-label-pos=3 --color=label:italic:black

# Right-aligned (negative integer)
fzf --height=10 --border-label="╢ $label ╟" --border=bottom --border-label-pos=-3 --color=label:italic:black

```

<img width="561" alt="image" src="https://user-images.githubusercontent.com/700826/198839268-f46d89d7-67d4-4db7-b9ac-094ea8d8bbda.png">

<img width="561" alt="image" src="https://user-images.githubusercontent.com/700826/198839279-83729770-fa8f-4759-9550-72676a383a67.png">

<img width="561" alt="image" src="https://user-images.githubusercontent.com/700826/198839294-5788abf5-f037-4dd8-8c14-3323194a1480.png">

**Comment by BlueFalconHD** [comment]
@junegunn did you actually implement the code yet? I am waiting for an awesome PR! Thanks so much for doing this, as I have no idea how to use Golang, and i'm a little busy to learn it at the moment.

**Comment by junegunn** [comment]
Yeah, I just need to write some tests and review the implementation.

**Comment by mhanberg** [comment]
What are your thoughts on allowing the border label to work with the index expressions (I'm forgetting the exact terminology)?

I use fzf with LSP workspaces symbols, references, etc and would love to be able to make the border label the file that is currently being displayed in the preview window. 

Main desire is to not have the file name being matched on, hide it in the list of results, but show file name in the border label.

**Comment by junegunn** [comment]
@mhanberg Can you show me a screenshot?

**Comment by mhanberg** [comment]
![fzf-exmaple](https://user-images.githubusercontent.com/5523984/198892025-a100c854-acfe-4db2-872d-30c3ed8e6490.png)

@junegunn Yes, sorry about that.

Here you can on each line, I have the ellipsis, which is hiding the file name, as I don't want to be matching on that as I type. (labeled with a red B)

But what I would be happy to see is the ability to put the filename in the border title, which I have mocked up at next to the red label, A. I don't necessarily _need_ the border title to be on the preview border, putting on the right side of the main border would be sufficient. 

I _think_ that the syntax it would probably use would be `--border-label="╢ {-1} ╟"`

**Comment by junegunn** [comment]
@mhanberg That's the border of the preview window, not the border of the main window. So it won't be covered by `--border-label`, but `--preview-label` which I also have a prototype implementation locally, albeit not dynamic.

However, after toying around with it for a little while, I think in most cases, printing the title inside the preview window is good enough and looks better. Have you tried something like this?

```sh
fzf --preview 'bat --color=always {} --highlight-line 50' --preview-window '~3,+50+3/2'

ag . | fzf -d: --preview 'bat --color=always --highlight-line {2} {1}' --preview-window '~3,+{2}+3/2'
```

**Comment by junegunn** [comment]
While making the label dynamic is doable, it involves more code for extra work (have to repeatedly re-evaluate the label expression, parse the ANSI color codes in it, and redraw it on the border). I'm not sure yet if it's worth the effort.

**Comment by mhanberg** [comment]
@junegunn turns out the plugin i was using explicitly sets bat to only have line numbers (no filename in the header), but i was able to get it configured correctly for me.

Thanks a lot!
